sync/statesync/code_queue.go

```diff
@@ -180,13 +181,14 @@ func (q *CodeQueue) enqueue(codeHashes []common.Hash) error {
 // Waits for in-flight enqueues to complete, then closes the output channel.
 // If the queue was already closed due to early quit, returns errFailedToFinalizeCodeQueue.
 func (q *CodeQueue) Finalize() error {
-	q.enqueueWG.Wait()
-	// If already closed due to quit, report early-exit error.
-	if q.closed.get() == closeStateQuit {
+	// Attempt to transition to finalized. If already quit, report early-exit error.
+	if q.closed.get() == closeStateQuit || !q.closed.canTransitionToFinalized() {
```

Comment:
```suggestion
	if !q.closed.canTransitionToFinalized() {
```

The check is redundant because the CaS guarantees that it's in open state.